From 9e563cc43ebae892a7332256e57586b9a96c21b3 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Tue, 1 Mar 2022 17:26:33 +0800
Subject: [PATCH 3530/4585] [Huawei] mm: Count reliable memory info based on
 zone info

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187270

--------------------------------

Count reliable memory info based on zone info. Any zone below
ZONE_MOVABLE is seed as reliable zone and sum the pages there.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 Documentation/filesystems/proc.rst |  3 ++
 include/linux/mem_reliable.h       |  6 ----
 include/linux/mm.h                 |  6 ++--
 mm/mem_reliable.c                  | 51 ++++++++++--------------------
 4 files changed, 23 insertions(+), 43 deletions(-)

diff --git a/Documentation/filesystems/proc.rst b/Documentation/filesystems/proc.rst
index f6783bb99e3f..6ae531ee4de9 100644
--- a/Documentation/filesystems/proc.rst
+++ b/Documentation/filesystems/proc.rst
@@ -971,6 +971,7 @@ varies by architecture and compile options.  The following is from a
     ShmemPmdMapped:      0 kB
     ReliableTotal: 7340032 kB
     ReliableUsed:   418824 kB
+    ReliableBuddyMem: 418824 kB
 
 MemTotal
               Total usable RAM (i.e. physical RAM minus a few reserved
@@ -1104,6 +1105,8 @@ ReliableTotal
               Total reliable memory size
 ReliableUsed
               The used amount of reliable memory
+ReliableBuddyMem
+              Size of unused mirrored memory in buddy system
 
 vmallocinfo
 ~~~~~~~~~~~
diff --git a/include/linux/mem_reliable.h b/include/linux/mem_reliable.h
index b75feac5e33c..dbe898a518b0 100644
--- a/include/linux/mem_reliable.h
+++ b/include/linux/mem_reliable.h
@@ -28,11 +28,6 @@ static inline bool mem_reliable_is_enabled(void)
 	return static_branch_likely(&mem_reliable);
 }
 
-static inline bool zone_reliable(struct zone *zone)
-{
-	return mem_reliable_is_enabled() && zone_idx(zone) < ZONE_MOVABLE;
-}
-
 static inline bool skip_none_movable_zone(gfp_t gfp, struct zoneref *z)
 {
 	if (!mem_reliable_is_enabled())
@@ -63,7 +58,6 @@ static inline void add_reliable_mem_size(long sz) {}
 static inline void mem_reliable_init(bool has_unmirrored_mem,
 				     unsigned long *zone_movable_pfn) {}
 static inline void shmem_reliable_init(void) {}
-static inline bool zone_reliable(struct zone *zone) { return false; }
 static inline bool skip_none_movable_zone(gfp_t gfp, struct zoneref *z)
 {
 	return false;
diff --git a/include/linux/mm.h b/include/linux/mm.h
index b1c3f72fa1b7..1e2c4a749a8d 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -34,9 +34,6 @@
 #include <linux/pgtable.h>
 #include <linux/kabi.h>
 
-/* added to mm.h to avoid every caller adding new header file */
-#include <linux/mem_reliable.h>
-
 struct mempolicy;
 struct anon_vma;
 struct anon_vma_chain;
@@ -3266,5 +3263,8 @@ static inline int seal_check_future_write(int seals, struct vm_area_struct *vma)
 	return 0;
 }
 
+/* added to mm.h to avoid every caller adding new header file */
+#include <linux/mem_reliable.h>
+
 #endif /* __KERNEL__ */
 #endif /* _LINUX_MM_H */
diff --git a/mm/mem_reliable.c b/mm/mem_reliable.c
index d46fe86563bd..ec58949f4af3 100644
--- a/mm/mem_reliable.c
+++ b/mm/mem_reliable.c
@@ -38,45 +38,34 @@ static unsigned long total_reliable_mem_sz(void)
 	return atomic_long_read(&total_reliable_mem);
 }
 
-static unsigned long used_reliable_mem_sz(void)
+static unsigned long total_reliable_pages(void)
 {
-	unsigned long nr_page = 0;
+	unsigned long total_reliable_pages = 0;
 	struct zone *z;
 
 	for_each_populated_zone(z)
 		if (zone_idx(z) < ZONE_MOVABLE)
-			nr_page += zone_page_state(z, NR_FREE_PAGES);
+			total_reliable_pages += zone_managed_pages(z);
 
-	return total_reliable_mem_sz() - nr_page * PAGE_SIZE;
+	return total_reliable_pages;
 }
 
-static int reliable_mem_notifier(struct notifier_block *nb,
-				 unsigned long action, void *arg)
+static unsigned long free_reliable_pages(void)
 {
-	struct memory_notify *m_arg = arg;
 	struct zone *zone;
+	unsigned long cnt = 0;
 
-	switch (action) {
-	case MEM_ONLINE:
-		zone = page_zone(pfn_to_page(m_arg->start_pfn));
-		if (zone_reliable(zone))
-			add_reliable_mem_size(m_arg->nr_pages * PAGE_SIZE);
-		break;
-	case MEM_OFFLINE:
-		zone = page_zone(pfn_to_page(m_arg->start_pfn));
-		if (zone_reliable(zone))
-			add_reliable_mem_size(-m_arg->nr_pages * PAGE_SIZE);
-		break;
-	default:
-		break;
-	}
+	for_each_populated_zone(zone)
+		if (zone_idx(zone) < ZONE_MOVABLE)
+			cnt += zone_page_state(zone, NR_FREE_PAGES);
 
-	return NOTIFY_OK;
+	return cnt;
 }
 
-static struct notifier_block reliable_notifier_block = {
-	.notifier_call = reliable_mem_notifier,
-};
+static unsigned long used_reliable_pages(void)
+{
+	return total_reliable_pages() - free_reliable_pages();
+}
 
 void mem_reliable_init(bool has_unmirrored_mem, unsigned long *zone_movable_pfn)
 {
@@ -95,11 +84,6 @@ void mem_reliable_init(bool has_unmirrored_mem, unsigned long *zone_movable_pfn)
 		return;
 	}
 
-	if (register_hotmemory_notifier(&reliable_notifier_block)) {
-		pr_err("init failed, register memory notifier failed.\n");
-		return;
-	}
-
 	static_branch_enable(&mem_reliable);
 
 	pr_info("init succeed, mirrored memory size(%lu)\n",
@@ -123,8 +107,7 @@ void reliable_report_meminfo(struct seq_file *m)
 	if (!mem_reliable_is_enabled())
 		return;
 
-	show_val_kb(m, "ReliableTotal:    ",
-			total_reliable_mem_sz() >> PAGE_SHIFT);
-	show_val_kb(m, "ReliableUsed:     ",
-			used_reliable_mem_sz() >> PAGE_SHIFT);
+	show_val_kb(m, "ReliableTotal:    ", total_reliable_pages());
+	show_val_kb(m, "ReliableUsed:     ", used_reliable_pages());
+	show_val_kb(m, "ReliableBuddyMem: ", free_reliable_pages());
 }
-- 
2.17.1

